### PR TITLE
filter: Add scheme field to cache keys and deprecate clear_http

### PR DESCRIPTION
**Pull request by edin44**
Signed-off-by: Edin Kadric <edin@google.com>

Commit Message: Add scheme field to Envoy.Extensions.HttpFilters.Cache.Key and deprecate clear_http. The current state is confusing because the clear_http field is not marked "optional", so it doesn't have a "has" presence test, which means that if protocol information is ignored and the field isn't set, users of the proto will interpret clear_http=false as SSL=true.
Additional Description: An alternative would be to just add "optional" to the field, but the field is also confusing because "clear_http" sounds like an action where we "clear" an HTTP field. Also, adding presence tracking to an existing field may not be safe: If the field is set to its default value and passes through an old binary, the field will be dropped and any newer binaries downstream will think the field is unset.
Risk Level: Low
Testing: N/A
Docs Changes: N/A
Release Notes: N/A (the CacheFilter is still in Alpha)
Platform Specific Features: N/A

### CODE REVIEW

**Comment by jmarantz** [comment]
@adisuissa for a first pass; actually this is pretty simple and can go straight to a non-google senior maintainer next.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Technically it seems this proto file is not serialized (other than for computing its hash), and is only used internally, so you may consider replacing the field instead of deprecating it.

Also can you please provide some details on how an http/s request have no scheme?

File: source/extensions/filters/http/cache/key.proto
Comment (edin44) [reply source/extensions/filters/http/cache/key.proto:18]: Done.

File: source/extensions/filters/http/cache/key.proto
Comment (edin44) [reply source/extensions/filters/http/cache/key.proto:12]: I think it would be reasonable to give it a month or two.

**Comment by edin44** [comment]
> Technically it seems this proto file is not serialized (other than for computing its hash), and is only used internally, so you may consider replacing the field instead of deprecating it.

This proto is part of the interface with the CacheFilter (through HttpCache and LookupRequest), so if anyone wrote a plugin for the CacheFilter, they would be using it, potentially hashing it to use as a cache key. Unfortunately we don't have a way to know that, so I'd rather play it safe and add a new field.

> Also can you please provide some details on how an http/s request have no scheme?

CDNs typically allow customers to configure different caching policies, with one of the parameters being whether a URL should map to the same cache entry for http vs https, or whether those should be separate cache entries. So the request will be either http or https, but as far as the cache key is concerned, we may choose to ignore that information.

**Comment by adisuissa** [comment]
> This proto is part of the interface with the CacheFilter (through HttpCache and LookupRequest), so if anyone wrote a plugin for the CacheFilter, they would be using it, potentially hashing it to use as a cache key. Unfortunately we don't have a way to know that, so I'd rather play it safe and add a new field.

I see, so to be backwards compatible, do you think the default enum value should be `HTTPS` (the default value of `clear_http` -> false)?

**Comment by edin44** [comment]
> I see, so to be backwards compatible, do you think the default enum value should be HTTPS (the default value of clear_http -> false)?

I don't think so, because this is a new field, so its future users can choose to interpret UNSPECIFIED as HTTPS when trying to be backwards compatible.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

**Review by toddmgreer — APPROVED** [review]

File: source/extensions/filters/http/cache/key.proto
Comment (alyssawilk) [inline source/extensions/filters/http/cache/key.proto:18]: Should this have a comment?  Especially what the behavior is for unspecified, which was not previously an option.

File: source/extensions/filters/http/cache/key.proto
Comment (alyssawilk) [reply source/extensions/filters/http/cache/key.proto:12]: I don't think we should deprecate until the new one actually works?  Also should we mark the new one as // [#not-implemented-hide:] to indicate it is not yet implemented?  Generally I encourage API changes to land with the code.  WDYT about delaying this until you've got the code prepared?

**Review by alyssawilk — COMMENTED** [review]
Thanks for picking this up!  Couple of comments

File: source/extensions/filters/http/cache/key.proto
Comment (edin44) [reply source/extensions/filters/http/cache/key.proto:18]: Done.

File: source/extensions/filters/http/cache/key.proto
Comment (edin44) [reply source/extensions/filters/http/cache/key.proto:12]: Oh, good point! I thought clear_http wasn't used yet but there is one file where it's used (http_cache.cc). I've made the change there. I'll also see if I can add test coverage (I'll need another commit in this PR).

This brings up the question of whether we need to protect this change more. Since this PR also changes the code that creates the cache key, then it could cause a cache flush for anyone who's using the CacheFilter already. The CacheFilter is still in Alpha, so maybe it's ok? The one example committed to the repo (SimpleHttpCache) is based on an in-memory cache which gets flushed on restart anyway.

Comment (edin44) [reply source/extensions/filters/http/cache/key.proto:12]: Added tests. Note that the UNSPECIFIED option isn't used currently. It will be used once we refactor LookupRequest::LookupRequest to call CachePolicy::createCacheKey to create the cache key. Then the cache policy can specify whether we want to include the scheme or not.

We've already made such a change internally and will work on upstreaming it in the coming weeks, along with many more changes, so I could wait until then and get this PR along with the rest of the relevant changes. However, I thought it would be a good idea to get this one in ASAP since it changes the structure of cache keys and will be more tricky to upstream the longer we wait.

And sorry for the messy PR, I'm still getting used to the Envoy Github workflow as a Googler...

Comment (alyssawilk) [reply source/extensions/filters/http/cache/key.proto:12]: yeah I think with the code changes added we can in-place deprecate as long as we have sign-off from one of the cache owners.  I'll add them in to the review

**Comment by alyssawilk** [comment]
cc @mpwarres @toddmgreer

**Comment by mpwarres** [comment]
LGTM. This is the upstreaming of a change already made in our internal version that @toddmgreer and I have already reviewed and approved. Also cc @capoferro

**Review by alyssawilk — APPROVED** [review]
Oh excellent, if it's been cleared by owners, LGTM :-)

**Pull request merged**
